clearwater-infrastructure/usr/share/clearwater/infrastructure/scripts/logging.bash

```diff
@@ -0,0 +1,174 @@
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+# This module is designed to provide a common logging facility which may be
+# used by all Bash scripts in Clearwater. It is intended to be sourced by other
+# scripts, passing in the following arguments:
+#   $1 - The name of the log file to create.
+#   $2 - The logging level. Defaults to WARNING
+#   $3 - The total number of log files to keep. Defaults to 1.
+
+function _logging_init()
+{
+  local filename="$1"
+  local log_level="$2"
+  local num_files="$3"
+
+  if ! [[ "$num_files" =~ ^[0-9]+$ ]]
+  then
+    echo "ERROR: Number of log backups is not a valid integer:" $num_files
+    return 1
+  fi
+
+  if (( "$num_files" > 1 ))
+  then
+    # Create a temporary logrotate config file to rotate old log files.
+    local full_filename=$(readlink -f "$filename")
+    local config_filename=$full_filename.logging.cfg
+    cat <<EOF > $config_filename
+$full_filename {
+  rotate $num_files
+  nocompress
+  missingok
+}
+EOF
+
+    # The default state file isn't writable by all users, so create a temp one
+    # here and delete it when we're done.
+    logrotate -f $config_filename -s $full_filename.logging.state
+    rm $config_filename
+    rm $full_filename.logging.state
+  fi
+
+  # Prevent namespace pollution
+  local file=""
+  local index=1
+
+  # logrotate doesn't look for log files with numbers greater than the value of
+  # the -c parameter + 1 ($num_files) so we should clean these up manually
+  for file in $filename.*
+  do
+    # Strip the filename base to get the index.
+    index="${file##$filename\.}"
+
+    if  ! [[ "$index" =~ ^[0-9]+$ ]]
+    then
+      continue
+    fi
+
+    # >= here as we want to leave exactly $num_files, including $filename, and
+    # our indexing of backup files starts at 1.
+    if (( $index >= $num_files ))
+    then
+      rm "$file"
+    fi
+  done
+
+  truncate --size 0 "$filename"
+
+  # Convert the log_level into an integer for convenience. Also check that the
+  # log level is valid. Note that these integers must match those used in the
+  # log_* functions below.
+  _logging_log_level=$(_logging_get_lvl_int "$log_level")
+
+  if [[ $? -ne 0 ]]
+  then
+    echo "ERROR: Invalid logging level given:" $log_level
+    return 1
+  fi
+}
+
+function _logging_get_lvl_int() {
+  # Convert the log_level into an integer for convenience. Also check that the
+  # log level is valid. Note that more severe log levels must have higher
+  # integers. This must be run in a sub-shell.
```

Comment:
I think the only reason for running in a subshell is the `exit` below. You could use return instead. This will set the subshell's exit code to the return value, and doesn't restrict you to using a subshell.